src/util/src/algebra_utils.rs

```diff
@@ -0,0 +1,499 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+
+#![deny(clippy::unwrap_used)]
+#![deny(clippy::expect_used)]
+#![deny(clippy::panic)]
+#![deny(clippy::manual_assert)]
+
+//! This module provides various utility functions for field and group elements.
+
+use itertools::Itertools;
+use std::{borrow::Borrow, collections::HashMap, iter::zip, mem};
+
+use num_bigint::{BigInt, BigUint, Sign};
+use num_integer::Integer;
+use num_traits::{One, Zero};
+
+use crate::algebra::{FieldElement, Group, GroupElement, ScalarField};
+
+/// Returns the number of bits required to encode the given number.
+pub fn cnt_bits_repr_usize(n: usize) -> usize {
+    if n == 0 {
+        1
+    } else {
+        n.ilog2() as usize + 1
+    }
+}
+
+/// Returns the number of bits required to encode the given number.
+pub fn cnt_bits_repr<T: Borrow<BigUint>>(n: &T) -> usize {
+    let n: &BigUint = n.borrow();
+    if n.is_zero() {
+        1
+    } else {
+        n.bits() as usize
+    }
+}
+
+pub fn largest_integer_a_such_that_2_to_a_divides_even_n(n: &BigUint) -> u64 {
+    assert!(n.is_even(), "requires n even");
+    assert!(!n.is_zero(), "requires n > 1");
+
+    // `unwrap()` is justified here because we just verified that 2 <= `n`.
+    #[allow(clippy::unwrap_used)]
+    n.trailing_zeros().unwrap()
+}
+
+// Returns the smallest integer `b` such that `b > a && b | n`.
+pub fn round_to_next_multiple(a: usize, x: usize) -> usize {
+    if a % x == 0 {
+        a
+    } else {
+        a + (x - (a % x))
```

Comment:
So this can overflow. Is that fine?

Reply:
The function is from existing code, but not used. I have removed it.